File: everything_parsing/26-table-driven-parsing/js/main.mjs
// node js/main.mjs   — identical output bytes to the Java build.

import { readFileSync } from "node:fs";
import { Grammar } from "./grammar.mjs";
import { TableParser, compact, render } from "./tableparser.mjs";

let sb = "";
const p = (s = "") => { sb += s + "\n"; };

p("=== Module 26 - Table-Driven Predictive Parsing ===");
p();
p("stack + LL(1) table + a loop. no recursion, no per-grammar code.");
p("stack top = nonterminal -> M[top][lookahead] gives the production; pop, push RHS reversed.");
p("stack top = terminal    -> must equal lookahead; pop and advance.");
p("stack top = $ and lookahead = $ -> ACCEPT.");
p();

const g = Grammar.parse(readFileSync("fixtures/expr.grammar", "utf8"));
p("grammar:");
for (const prod of g.productions) p("  " + compact(prod));
p();

const parser = new TableParser(g);

for (const line of readFileSync("fixtures/inputs.txt", "utf8").split(/\r?\n/)) {
  if (line.trim() === "" || line.startsWith("#")) continue;
  const toks = line.trim().split(/\s+/);

  p("--- " + line + " ---");
  const r = parser.parse(toks);
  for (const t of r.trace) p(t);

  if (r.ok) {
    p("  productions (a leftmost derivation):");
    for (const prod of r.productions) p("      " + prod);
    const sform = parser.replay(r.productions);
    p("  replay -> " + sform.join(" "));
    p("  equals input? " + (sform.length === toks.length && sform.every((x, i) => x === toks[i])));
    p("  parse tree:");
    for (const tl of render(r.tree).split("\n")) p("    " + tl);
  } else {
    p("  REJECTED: " + r.error);
  }
  p();
}

p("table-driven == recursive descent: the production sequence above is the");
p("same leftmost derivation Module 19's call stack produces for the same input.");
p("the explicit stack here IS that call stack, made into data.");

process.stdout.write(sb);

File: everything_parsing/26-table-driven-parsing/java/Main.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;

public final class Main {

    public static void main(String[] args) throws Exception {
        StringBuilder sb = new StringBuilder();
        sb.append("=== Module 26 - Table-Driven Predictive Parsing ===\n\n");
        sb.append("stack + LL(1) table + a loop. no recursion, no per-grammar code.\n");
        sb.append("stack top = nonterminal -> M[top][lookahead] gives the production; pop, push RHS reversed.\n");
        sb.append("stack top = terminal    -> must equal lookahead; pop and advance.\n");
        sb.append("stack top = $ and lookahead = $ -> ACCEPT.\n\n");

        Grammar g = Grammar.parse(Files.readString(Path.of("fixtures/expr.grammar")));
        sb.append("grammar:\n");
        for (Grammar.Production p : g.productions) sb.append("  ").append(TableParser.compact(p)).append('\n');
        sb.append('\n');

        TableParser parser = new TableParser(g);

        for (String line : Files.readAllLines(Path.of("fixtures/inputs.txt"))) {
            if (line.isBlank() || line.startsWith("#")) continue;
            List<String> toks = Arrays.asList(line.trim().split("\\s+"));

            sb.append("--- ").append(line).append(" ---\n");
            TableParser.Result r = parser.parse(toks);

            for (String t : r.trace()) sb.append(t).append('\n');

            if (r.ok()) {
                sb.append("  productions (a leftmost derivation):\n");
                for (String prod : r.productions()) sb.append("      ").append(prod).append('\n');
                List<String> sform = parser.replay(r.productions());
                sb.append("  replay -> ").append(String.join(" ", sform)).append('\n');
                sb.append("  equals input? ").append(sform.equals(toks)).append('\n');
                sb.append("  parse tree:\n");
                for (String tl : TableParser.render(r.tree()).split("\n"))
                    sb.append("    ").append(tl).append('\n');
            } else {
                sb.append("  REJECTED: ").append(r.error()).append('\n');
            }
            sb.append('\n');
        }

        sb.append("table-driven == recursive descent: the production sequence above is the\n");
        sb.append("same leftmost derivation Module 19's call stack produces for the same input.\n");
        sb.append("the explicit stack here IS that call stack, made into data.\n");

        System.out.print(sb.toString().replace("\r\n", "\n"));
    }
}
